File: app/src/main/java/top/easelink/lcg/ui/main/articles/view/ForumArticlesFragment.java
package top.easelink.lcg.ui.main.articles.view;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.tabs.TabLayout;
import com.tencent.stat.StatService;

import java.util.List;

import javax.inject.Inject;

import top.easelink.framework.base.BaseFragment;
import top.easelink.framework.customview.ScrollChildSwipeRefreshLayout;
import top.easelink.lcg.BR;
import top.easelink.lcg.LCGApp;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.FragmentForumArticlesBinding;
import top.easelink.lcg.ui.ViewModelProviderFactory;
import top.easelink.lcg.ui.main.articles.viewmodel.ArticlesAdapter;
import top.easelink.lcg.ui.main.articles.viewmodel.ForumArticlesViewModel;
import top.easelink.lcg.ui.main.model.LoginRequiredException;
import top.easelink.lcg.ui.main.source.model.ForumThread;

import static top.easelink.lcg.mta.MTAConstantKt.CHANGE_THREAD;
import static top.easelink.lcg.ui.main.articles.viewmodel.ArticlesAdapter.ArticlesAdapterListener.FETCH_BY_THREAD;
import static top.easelink.lcg.ui.main.articles.viewmodel.ArticlesViewModel.FETCH_INIT;

public class ForumArticlesFragment extends BaseFragment<FragmentForumArticlesBinding, ForumArticlesViewModel>
        implements FourmArticleNavigator {

    private static final String ARG_PARAM = "url";
    private static final String ARG_TITLE = "title";

    @Inject
    ViewModelProviderFactory factory;
    private FragmentForumArticlesBinding mFragmentArticlesBinding;
    private ForumArticlesViewModel mArticlesViewModel;

    public static ForumArticlesFragment newInstance(String title, String param) {
        Bundle args = new Bundle();
        args.putString(ARG_PARAM, param);
        args.putString(ARG_TITLE, title);
        ForumArticlesFragment fragment = new ForumArticlesFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_forum_articles;
    }

    @Override
    public ForumArticlesViewModel getViewModel() {
        mArticlesViewModel = ViewModelProviders.of(this, factory).get(ForumArticlesViewModel.class);
        return mArticlesViewModel;
    }

    @Override
    public void handleError(Throwable t) {
        int errorId;
        if (t instanceof LoginRequiredException) {
            errorId = R.string.login_required_error;
        } else {
            errorId = R.string.error;
        }
        showMessage(errorId);
        mFragmentArticlesBinding.loading.setAnimation(R.raw.face_sorry);
    }

    @Override
    public void scrollToTop() {
        mFragmentArticlesBinding.recyclerView.smoothScrollToPosition(0);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mArticlesViewModel.setNavigator(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mFragmentArticlesBinding = getViewDataBinding();
        setUp();
    }

    private void setUp() {
        Bundle bundle = getArguments();
        if (bundle != null) {
            mArticlesViewModel.initUrl(getArguments().getString(ARG_PARAM), FETCH_INIT);
            mArticlesViewModel.setTitle(getArguments().getString(ARG_TITLE));
        }

        setUpRecyclerView();
    }

    @Override
    public void setUpTabLayout(List<ForumThread> forumThreadList) {
        TabLayout tabLayout = mFragmentArticlesBinding.forumTab;
        tabLayout.setVisibility(View.VISIBLE);
        for(ForumThread forumThread: forumThreadList) {
            tabLayout.addTab(tabLayout.newTab().setText(forumThread.getThreadName()));
        }

        mFragmentArticlesBinding.forumTab.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                StatService.trackCustomEvent(getContext(), CHANGE_THREAD);
                mArticlesViewModel.initUrl(
                        forumThreadList.get(tab.getPosition()).getThreadUrl(),
                        FETCH_BY_THREAD);
            }
            @Override
            public void onTabUnselected(TabLayout.Tab tab) { }
            @Override
            public void onTabReselected(TabLayout.Tab tab) { }
        });
    }

    private void setUpRecyclerView() {
        ArticlesAdapter mArticlesAdapter = new ArticlesAdapter(mArticlesViewModel);
        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getContext());
        mLayoutManager.setOrientation(RecyclerView.VERTICAL);
        mFragmentArticlesBinding.recyclerView.setLayoutManager(mLayoutManager);
        mFragmentArticlesBinding.recyclerView.setItemAnimator(new DefaultItemAnimator());
        mFragmentArticlesBinding.recyclerView.setAdapter(mArticlesAdapter);
        final ScrollChildSwipeRefreshLayout swipeRefreshLayout = getViewDataBinding().refreshLayout;
        Context context = LCGApp.getContext();
        swipeRefreshLayout.setColorSchemeColors(
                ContextCompat.getColor(context, R.color.colorPrimary),
                ContextCompat.getColor(context, R.color.colorAccent),
                ContextCompat.getColor(context, R.color.colorPrimaryDark)
        );
        // Set the scrolling view in the custom SwipeRefreshLayout.
        swipeRefreshLayout.setScrollUpChild(mFragmentArticlesBinding.recyclerView);
        swipeRefreshLayout.setOnRefreshListener(() -> mArticlesViewModel.fetchArticles(FETCH_INIT));
    }

    @Override
    public void showMessage(@StringRes int resId) {
        Toast.makeText(getContext(), resId, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/articles/view/ForumArticlesFragment.kt
package top.easelink.lcg.ui.main.articles.view

import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayout.OnTabSelectedListener
import com.tencent.stat.StatService
import kotlinx.android.synthetic.main.fragment_forum_articles.*
import timber.log.Timber
import top.easelink.framework.base.BaseFragment
import top.easelink.lcg.BR
import top.easelink.lcg.LCGApp
import top.easelink.lcg.R
import top.easelink.lcg.databinding.FragmentForumArticlesBinding
import top.easelink.lcg.mta.CHANGE_THREAD
import top.easelink.lcg.ui.main.articles.viewmodel.*
import top.easelink.lcg.ui.main.source.model.ForumThread

class ForumArticlesFragment : BaseFragment<FragmentForumArticlesBinding, ForumArticlesViewModel>() {

    private var showTab = false

    override fun isControllable(): Boolean {
        return true
    }

    override fun getBindingVariable(): Int {
        return BR.viewModel
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_forum_articles
    }

    override fun getViewModel(): ForumArticlesViewModel {
        return ViewModelProviders.of(this).get(ForumArticlesViewModel::class.java)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        arguments?.run {
            showTab = getBoolean(ARG_SHOW_TAB, true)
        }
        setUpToolbar()
        setUp()
    }

    private fun setUpToolbar() {
        article_toolbar.apply {
            inflateMenu(R.menu.forum_articles)
            setOnMenuItemClickListener { menuItem ->
                val order = when (menuItem.itemId) {
                    R.id.action_order_by_datetime -> DATE_LINE_ORDER
                    R.id.action_order_by_lastpost -> LAST_POST_ORDER
                    else -> DEFAULT_ORDER
                }
                try {
                    val pos = viewDataBinding.forumTab.selectedTabPosition
                    viewModel
                        .threadList
                        .value
                        ?.get(pos)
                        ?.threadUrl
                        ?.let {
                            viewModel.initUrlAndFetch(
                                url = it,
                                fetchType = ArticleFetcher.FetchType.FETCH_INIT,
                                order = order
                            )
                        }
                } catch (e: Exception) {
                    Timber.e(e)
                }
                return@setOnMenuItemClickListener true
            }
        }
    }


    private fun setUp() {
        viewModel.threadList.observe(viewLifecycleOwner, Observer {
                threadList -> setUpTabLayout(threadList)
        })
        arguments?.let {
            viewModel.initUrlAndFetch(
                url = it.getString(ARG_PARAM)!!,
                fetchType = ArticleFetcher.FetchType.FETCH_INIT)
            viewModel.setTitle(it.getString(ARG_TITLE).orEmpty())
        }
        setUpRecyclerView()
    }

    private fun setUpTabLayout(forumThreadList: List<ForumThread>?) {
        // hide tab if showTab has been set to false or there's no tab item
        if (forumThreadList.isNullOrEmpty() || !showTab) {
            viewDataBinding.forumTab.apply {
                visibility = View.GONE
                removeAllTabs()
            }
            return
        }
        viewDataBinding.forumTab.apply {
            visibility = View.VISIBLE
            removeAllTabs()
            forumThreadList.forEach {
                addTab(newTab().setText(it.threadName))
            }
            addOnTabSelectedListener(object : OnTabSelectedListener {
                override fun onTabSelected(tab: TabLayout.Tab) {
                    StatService.trackCustomEvent(context, CHANGE_THREAD)
                    viewModel.initUrlAndFetch(
                        url = forumThreadList[tab.position].threadUrl,
                        fetchType = ArticleFetcher.FetchType.FETCH_INIT
                    )
                }
                override fun onTabUnselected(tab: TabLayout.Tab) {}
                override fun onTabReselected(tab: TabLayout.Tab) {}
            })
        }
    }

    private fun setUpRecyclerView() {
        viewDataBinding.refreshLayout.apply {
            LCGApp.context.let {
                setColorSchemeColors(
                    ContextCompat.getColor(it, R.color.colorPrimary),
                    ContextCompat.getColor(it, R.color.colorAccent),
                    ContextCompat.getColor(it, R.color.colorPrimaryDark)
                )
            }
            // Set the scrolling view in the custom SwipeRefreshLayout.
            setScrollUpChild(
                viewDataBinding.recyclerView.apply {
                    layoutManager = LinearLayoutManager(context).apply {
                        orientation = RecyclerView.VERTICAL
                    }
                    itemAnimator = DefaultItemAnimator()
                    adapter = ArticlesAdapter(viewModel).also {
                        it.setFragmentManager(fragmentManager?:baseActivity.supportFragmentManager)
                    }
                }
            )
            setOnRefreshListener {
                viewModel.fetchArticles(ArticleFetcher.FetchType.FETCH_INIT){}
            }
        }
        // Add articles observer
        viewModel.articles.observe(viewLifecycleOwner, Observer { articleList ->
            if (articleList.isEmpty() && viewModel.isLoading.value == true) {
                viewDataBinding.recyclerView.visibility = View.GONE
            } else {
                viewDataBinding.recyclerView.visibility = View.VISIBLE
                (viewDataBinding.recyclerView.adapter as? ArticlesAdapter)?.apply {
                    // a workaround for distinguish fetch_init / fetch_more / thread
                    clearItems()
                    addItems(articleList)
                }
            }
        })
    }

    companion object {
        private const val ARG_PARAM = "url"
        private const val ARG_TITLE = "title"
        private const val ARG_SHOW_TAB = "showTab"
        @JvmStatic
        fun newInstance(title: String, param: String, showTab: Boolean = true): ForumArticlesFragment {
            val args = Bundle()
            args.putString(ARG_PARAM, param)
            args.putString(ARG_TITLE, title)
            args.putBoolean(ARG_SHOW_TAB, showTab)
            val fragment = ForumArticlesFragment()
            fragment.arguments = args
            return fragment
        }
    }
}